### PR TITLE
RetroAchievements CSO support, delay until identified

### PR DESCRIPTION
**Pull request by hrydgard**
Adds support for identifying CSO files for retroachievements.

Additionally, prevents auto-loading save states in challenge mode.

Also a correctness fix: Don't start executing the game until RetroAchievement identification is done.

See #17631

### CODE REVIEW

**Comment by hrydgard** [comment]
Oops, accidentally squash-merge, stupid github remembering what I did last. Oh well, just two commits.

File: Core/RetroAchievements.cpp [hunk 1]
```diff
@@ -605,22 +611,72 @@ void identify_and_load_callback(int result, const char *error_message, rc_client
...
+		FileContext *ctx = (FileContext *)file_handle;
+
+		int blockSize = ctx->bd->GetBlockSize();
+
+		int64_t offset = ctx->seekPos;
+		int64_t endOffset = ctx->seekPos + requested_bytes;
+		int firstBlock = offset / blockSize;
+		int afterLastBlock = (endOffset + blockSize - 1) / blockSize;
+		int numBlocks = afterLastBlock - firstBlock;
+		// This is suboptimal, but good enough since we're not doing a lot of accesses.
```
Comment (unknownbrackets) [inline Core/RetroAchievements.cpp:660]: Without looking super hard, it seems like it'd generally read aligned sectors:
https://github.com/RetroAchievements/rcheevos/blob/aa6324e886e12da0cf24dee1cf4e873571f39191/src/rhash/cdreader.c#L827C68-L827C81
https://github.com/RetroAchievements/rcheevos/blob/aa6324e886e12da0cf24dee1cf4e873571f39191/src/rhash/cdreader.c#L62

Overall for an entire ISO, might be worth checking if `offset == firstOffset && requested_bytes == numBlocks * blockSize` and just reading directly.  But probably the IO will dominate anyway...

-[Unknown]

Comment (hrydgard) [reply Core/RetroAchievements.cpp:660]: It only hashes a few files anyway, so there's no heavy pressure on this.

Also, the very first read it does is actually unaligned - I first tried to assume aligned sectors but that was a no-go, so I just made a general quickfix.

Comment (unknownbrackets) [reply Core/RetroAchievements.cpp:660]: Oh, it's not reading the whole ISO?  Then yeah, probably not a worry anyway.

-[Unknown]

Comment (hrydgard) [reply Core/RetroAchievements.cpp:660]: Yeah, it's not ideal but it's how RetroAchievements does things, and for user experience it's a good thing, of course... It hashes the eboot and paramsfo basically. Cheating through hacking other files is thus perfectly possible, but on the other hand, you can just hack the emulator instead...

**Pull request merged**